File: solutions/fable-1/fable/perl/dijkstra.pl
use strict;
use warnings;

my @tokens;
{
    local $/;
    my $data = <STDIN>;
    $data = '' unless defined $data;
    @tokens = split ' ', $data;
}
my $idx = 0;
my $n = $tokens[$idx++] // 0;
my $m = $tokens[$idx++] // 0;

my @adj;
$adj[$_] = [] for 0 .. $n - 1;
for my $i (1 .. $m) {
    my $u = $tokens[$idx++];
    my $v = $tokens[$idx++];
    my $w = $tokens[$idx++];
    last unless defined $w;
    push @{ $adj[$u] }, [ $v, $w ];
    push @{ $adj[$v] }, [ $u, $w ];
}
my $s = $tokens[$idx++] // 0;
my $t = $tokens[$idx++] // 0;

if ($s == $t) {
    print "0\n";
    exit 0;
}

my @dist;
$dist[$s] = 0;

# Binary min-heap of [dist, node]
my @heap = ( [ 0, $s ] );

sub heap_push {
    my ($item) = @_;
    push @heap, $item;
    my $i = $#heap;
    while ( $i > 0 ) {
        my $p = ( $i - 1 ) >> 1;
        last if $heap[$p][0] <= $heap[$i][0];
        @heap[ $p, $i ] = @heap[ $i, $p ];
        $i = $p;
    }
}

sub heap_pop {
    my $top  = $heap[0];
    my $last = pop @heap;
    if (@heap) {
        $heap[0] = $last;
        my $i = 0;
        my $sz = scalar @heap;
        while (1) {
            my $l = 2 * $i + 1;
            last if $l >= $sz;
            my $r = $l + 1;
            my $c = ( $r < $sz && $heap[$r][0] < $heap[$l][0] ) ? $r : $l;
            last if $heap[$i][0] <= $heap[$c][0];
            @heap[ $i, $c ] = @heap[ $c, $i ];
            $i = $c;
        }
    }
    return $top;
}

my $answer = -1;
while (@heap) {
    my ( $d, $u ) = @{ heap_pop() };
    next if $d > $dist[$u];
    if ( $u == $t ) {
        $answer = $d;
        last;
    }
    for my $e ( @{ $adj[$u] } ) {
        my ( $v, $w ) = @$e;
        my $nd = $d + $w;
        if ( !defined $dist[$v] || $nd < $dist[$v] ) {
            $dist[$v] = $nd;
            heap_push( [ $nd, $v ] );
        }
    }
}

print "$answer\n";

File: solutions/fable-1/fable/ruby/dijkstra.rb
# frozen_string_literal: true

# Binary min-heap over [dist, node] pairs.
class MinHeap
  def initialize
    @a = []
  end

  def empty?
    @a.empty?
  end

  def push(d, v)
    @a << [d, v]
    i = @a.size - 1
    while i > 0
      p = (i - 1) >> 1
      break if @a[p][0] <= @a[i][0]
      @a[p], @a[i] = @a[i], @a[p]
      i = p
    end
  end

  def pop
    top = @a[0]
    last = @a.pop
    unless @a.empty?
      @a[0] = last
      i = 0
      n = @a.size
      loop do
        l = 2 * i + 1
        r = l + 1
        m = i
        m = l if l < n && @a[l][0] < @a[m][0]
        m = r if r < n && @a[r][0] < @a[m][0]
        break if m == i
        @a[m], @a[i] = @a[i], @a[m]
        i = m
      end
    end
    top
  end
end

data = STDIN.read.split.map(&:to_i)
idx = 0
n = data[idx]; idx += 1
m = data[idx]; idx += 1

adj = Array.new(n) { [] }
m.times do
  u = data[idx]; v = data[idx + 1]; w = data[idx + 2]
  idx += 3
  adj[u] << [v, w]
  adj[v] << [u, w]
end

s = data[idx]; t = data[idx + 1]

if s == t
  puts 0
  exit
end

INF = Float::INFINITY
dist = Array.new(n, INF)
dist[s] = 0
heap = MinHeap.new
heap.push(0, s)
result = -1

until heap.empty?
  d, u = heap.pop
  next if d > dist[u]
  if u == t
    result = d
    break
  end
  adj[u].each do |v, w|
    nd = d + w
    if nd < dist[v]
      dist[v] = nd
      heap.push(nd, v)
    end
  end
end

puts result
